### PR TITLE
Add PriorityClass

### PR DESCRIPTION
**Pull request by Vincent056**
Added the ability to set `PriorityClass` through ScanSetting. All the scanner pods will be launched using the `PriorityClass` if set.

### CODE REVIEW

**Comment by xiaojiey** [comment]
@Vincent056 Is it ready for QE test now? Thanks.

**Comment by jhrozek** [comment]
/hold
because this is a feature and there is no clear way to test. I know that PR #50 introduced some docs, but do you think this is enough for QE to verify the feature before merging? 
Same for docs, we should get acks from QE and docs before merging.

**Comment by jhrozek** [comment]
btw I'll review the PR code-wise, but the point still stands

**Comment by Vincent056** [comment]
> 

I will add what's needed for QE and Doc

File: pkg/controller/compliancesuite/suitererunner.go [hunk 1]
```diff
@@ -98,9 +124,11 @@ func GetRerunnerName(suiteName string) string {
 	return suiteName + "-rerunner"
 }
 
-func (r *ReconcileComplianceSuite) getRerunner(suite *compv1alpha1.ComplianceSuite) *batchv1beta1.CronJob {
+func (r *ReconcileComplianceSuite) getRerunner(suite *compv1alpha1.ComplianceSuite) (*batchv1beta1.CronJob, error) {
 	falseP := false
 	trueP := true
+	priorityClassName, err := r.getPriorityClassName(suite)
```
[the remark this answers is not in the document]
Comment (Vincent056) [reply pkg/controller/compliancesuite/suitererunner.go:130]: As for right now, we are not validating if there is a priority class exists on the cluster, if the one reference is not there, it will just be ignored, when the pod is launched. I think I can add a warning if the priorityclass referenced was not there.

[the remark this answers is not in the document]
Comment (Vincent056) [reply pkg/controller/compliancesuite/suitererunner.go:130]: I am going to add validation for this

File: tests/e2e/e2e_test.go [hunk 2]
```diff
@@ -1802,6 +1805,90 @@ func TestE2E(t *testing.T) {
...
+							{
+								Name: workerScanName,
+								ComplianceScanSpec: compv1alpha1.ComplianceScanSpec{
+									ContentImage: contentImagePath,
+									Profile:      "xccdf_org.ssgproject.content_profile_moderate",
+									Content:      rhcosContentFile,
+									Rule:         "xccdf_org.ssgproject.content_rule_no_netrc_files",
+									NodeSelector: selectWorkers,
+									ComplianceScanSettings: compv1alpha1.ComplianceScanSettings{
+										PriorityClass: "e2e-compliance-suite-high-priority",
```
[the remark this answers is not in the document]
Comment (Vincent056) [reply tests/e2e/e2e_test.go:1851]: that's a good catch, it should be `test-scheduled-suite-priority-class`

File: tests/e2e/e2e_test.go [hunk 3]
```diff
@@ -1802,6 +1805,90 @@ func TestE2E(t *testing.T) {
...
+				}
+
+				// Ensure that all the scans in the suite have finished and are marked as Done
+				err = waitForSuiteScansStatus(t, f, namespace, suiteName, compv1alpha1.PhaseDone, compv1alpha1.ResultCompliant)
+				if err != nil {
+					return err
+				}
+				return nil
+			},
+		},
```
[the remark this answers is not in the document]
Comment (Vincent056) [reply tests/e2e/e2e_test.go:1891]: I will add that

**Comment by Vincent056** [comment]
@sheriff-rh  I have added the doc here: https://github.com/ComplianceAsCode/compliance-operator/pull/76,
please let me know if this is good for downstream or if more is needed, thanks!

**Comment by sheriff-rh** [comment]
Hey @Vincent056  thanks for bringing me in. 

I think we have enough information to refer to when adding this to downstream at a later date. Thanks!

File: pkg/controller/compliancesuite/suitererunner.go [hunk 4]
```diff
@@ -2,27 +2,39 @@ package compliancesuite
...
-	"github.com/openshift/compliance-operator/pkg/utils"
 )
 
 const rerunnerServiceAccount = "rerunner"
 
 func (r *ReconcileComplianceSuite) reconcileScanRerunnerCronJob(suite *compv1alpha1.ComplianceSuite, logger logr.Logger) error {
 	rerunner := r.getRerunner(suite)
+	priorityClassName, err := r.getPriorityClassName(suite)
+	if err != nil {
+		logger.Error(err, "Cannot get priority class name")
```
Comment (rhmdnd) [inline pkg/controller/compliancesuite/suitererunner.go:29]: We could also say here that the compliance operator is ignoring the priority class and the scan will not be run with a priority class.

File: pkg/controller/compliancesuite/suitererunner.go [hunk 5]
```diff
@@ -63,6 +75,26 @@ func (r *ReconcileComplianceSuite) handleCreate(suite *compv1alpha1.ComplianceSu
...
+		Namespace:     suite.Namespace,
+	}
+	err := r.client.List(context.TODO(), scans, listOpts)
+	if err != nil {
+		return "", fmt.Errorf("Error while getting scans for ComplianceSuite '%s', err: %s\n", suite.Name, err)
+	}
+	if len(scans.Items) == 0 {
+		return "", fmt.Errorf("No scans found for ComplianceSuite '%s'", suite.Name)
+	}
+	return scans.Items[0].Spec.PriorityClass, nil
```
Comment (rhmdnd) [inline pkg/controller/compliancesuite/suitererunner.go:95]: Is the `PriorityClass` attribute always present?

File: tests/e2e/helpers.go [hunk 6]
```diff
@@ -2082,6 +2093,50 @@ func waitForPod(podCallback wait.ConditionFunc) error {
...
+			E2ELogf(t, "pod %s has priority class %s, expected %s", podName, pod.Spec.PriorityClassName, priorityClass)
+			return true, nil
+		}
+
+		return true, nil
+	}
+}
+
+// check if pod has empty priority class
+func checkPodEmptyPriorityClass(t *testing.T, c kubernetes.Interface, podName, namespace string) wait.ConditionFunc {
```
Comment (rhmdnd) [inline tests/e2e/helpers.go:2119]: Couldn't we collapse these two assertions into a single function?

Do we need a separate function for checking empty pod priority classes if we have the ability to pass in "" as the priorityClass?

```go
// make sure the priority class matches one of the defaults
expectedPriorityClass := "system-cluster-critical"
waitForPod(checkPodPriorityClass(t, f.KubeClient, pod.Name, namespace, expectedPriorityClass))
// make sure the priority class is empty
expectedPriorityClass := ""
waitForPod(checkPodPriorityClass(t, f.KubeClient, pod.Name, namespace, expectedPriorityClass))
```

Might save us from having a separate, but nearly identical method.

File: CHANGELOG.md [hunk 7]
```diff
@@ -9,7 +9,8 @@ Versioning](https://semver.org/spec/v2.0.0.html).
 
 ### Enhancements
 
--
+- Added the ability to set `PriorityClass` through ScanSetting. All the scanner pods
```
Comment (rhmdnd) [inline CHANGELOG.md:12]: We could make `ScanSetting` a literal.

File: CHANGELOG.md [hunk 8]
```diff
@@ -9,7 +9,8 @@ Versioning](https://semver.org/spec/v2.0.0.html).
...
+  will be launched using the `PriorityClass` if set.
```
Comment (rhmdnd) [inline CHANGELOG.md:13]: We could also say this is an optional feature and that scans will continue to work.

File: deploy/compliance-operator-chart/crds/compliance.openshift.io_compliancescans_crd.yaml [hunk 9]
```diff
@@ -74,6 +74,11 @@ spec:
                   generated from the scan, this should match the selector of the MachineConfigPool
                   you want to apply the remediations to.
                 type: object
+              priorityClass:
+                description: Defines the PriorityClass to use for launching scan related
+                  pods, if PriorityClass is invalid and not found, we will treat scansetting
+                  didn't have a PriorityClass.
```
Comment (rhmdnd) [inline deploy/compliance-operator-chart/crds/compliance.openshift.io_compliancescans_crd.yaml:80]: After looking through the other description, I think we should avoid the first person point of view (e.g. removing "we").

Perhaps:

```yaml
description: Defines the PriorityClass to use for launching scan-related pods.
  If the PriorityClass is not found, pods won't be scheduled using a PriorityClass.
```

File: deploy/compliance-operator-chart/crds/compliance.openshift.io_compliancesuites_crd.yaml [hunk 10]
```diff
@@ -95,6 +95,11 @@ spec:
                         selector of the MachineConfigPool you want to apply the remediations
                         to.
                       type: object
+                    priorityClass:
+                      description: Defines the PriorityClass to use for launching
+                        scan related pods, if PriorityClass is invalid and not found,
+                        we will treat scansetting didn't have a PriorityClass.
+                      type: string
```
Comment (rhmdnd) [inline deploy/compliance-operator-chart/crds/compliance.openshift.io_compliancesuites_crd.yaml:102]: Similar comment as above.

File: deploy/compliance-operator-chart/crds/compliance.openshift.io_scansettings_crd.yaml [hunk 11]
```diff
@@ -52,6 +52,11 @@ spec:
               be used. External resources could be, for instance, CVE feeds. This
               is useful for disconnected installations without access to a proxy.
             type: boolean
+          priorityClass:
+            description: Defines the PriorityClass to use for launching scan related
+              pods, if PriorityClass is invalid and not found, we will treat scansetting
+              didn't have a PriorityClass.
+            type: string
```
Comment (rhmdnd) [inline deploy/compliance-operator-chart/crds/compliance.openshift.io_scansettings_crd.yaml:59]: Similar comment as above.

File: deploy/crds/compliance.openshift.io_compliancescans_crd.yaml [hunk 12]
```diff
@@ -74,6 +74,11 @@ spec:
                   generated from the scan, this should match the selector of the MachineConfigPool
                   you want to apply the remediations to.
                 type: object
+              priorityClass:
+                description: Defines the PriorityClass to use for launching scan related
+                  pods, if PriorityClass is invalid and not found, we will treat scansetting
+                  didn't have a PriorityClass.
+                type: string
```
Comment (rhmdnd) [inline deploy/crds/compliance.openshift.io_compliancescans_crd.yaml:81]: Similar comment as above.

File: deploy/crds/compliance.openshift.io_compliancesuites_crd.yaml [hunk 13]
```diff
@@ -95,6 +95,11 @@ spec:
                         selector of the MachineConfigPool you want to apply the remediations
                         to.
                       type: object
+                    priorityClass:
+                      description: Defines the PriorityClass to use for launching
+                        scan related pods, if PriorityClass is invalid and not found,
+                        we will treat scansetting didn't have a PriorityClass.
+                      type: string
```
Comment (rhmdnd) [inline deploy/crds/compliance.openshift.io_compliancesuites_crd.yaml:102]: Similar comment as above.

File: deploy/crds/compliance.openshift.io_scansettings_crd.yaml [hunk 14]
```diff
@@ -52,6 +52,11 @@ spec:
               be used. External resources could be, for instance, CVE feeds. This
               is useful for disconnected installations without access to a proxy.
             type: boolean
+          priorityClass:
+            description: Defines the PriorityClass to use for launching scan related
+              pods, if PriorityClass is invalid and not found, we will treat scansetting
+              didn't have a PriorityClass.
+            type: string
```
Comment (rhmdnd) [inline deploy/crds/compliance.openshift.io_scansettings_crd.yaml:59]: Similar comment as above.

File: doc/crds.md [hunk 15]
```diff
@@ -606,6 +606,8 @@ In the `spec`:
   for scheduling. Defaults to allowing the scan to run on master nodes. For
   details on tolerations, see the
   [Kubernetes documentation on this](https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/).
+* **PriorityClass**: Defines the priority class reference of the scan(s).
```
Comment (rhmdnd) [inline doc/crds.md:609]: Priority class reference, or name? Is there a difference?

**Review by rhmdnd — CHANGES_REQUESTED** [review]
Tested this out locally and it works well! Just some comments on wording and the opportunity to DRY up some of the test code.

Nice work, Vincent.

File: pkg/controller/compliancesuite/suitererunner.go (hunk 5 above)
Comment (Vincent056) [reply pkg/controller/compliancesuite/suitererunner.go:95]: It will be "", if `PriorityClass` is not set

File: doc/crds.md (hunk 15 above)
Comment (Vincent056) [reply doc/crds.md:609]: It should be priorityclass name

File: deploy/compliance-operator-chart/crds/compliance.openshift.io_compliancescans_crd.yaml [hunk 16]
```diff
@@ -74,6 +74,12 @@ spec:
                   generated from the scan, this should match the selector of the MachineConfigPool
                   you want to apply the remediations to.
                 type: object
+              priorityClass:
+                description: Defines the PriorityClass to use for launching scan related
+                  pods, the Name of a desired PriorityClass should be set here, this
+                  is an optional field, if PriorityClass is invalid, not found, or
+                  not set we will ignore this setting.
```
Comment (rhmdnd) [inline deploy/compliance-operator-chart/crds/compliance.openshift.io_compliancescans_crd.yaml:81]: I would reword this to remove "we", since we don't use first-person in other parts of the document from what I can tell.

Perhaps:

`this is an optional field, if PriorityClass is invalid or not found, it will be ignored.`

**Review by rhmdnd — COMMENTED** [review]
Looks good, just one minor wording comment.

**Review by rhmdnd — APPROVED** [review]
/lgtm

Thanks, Vincent!

**Comment by rhmdnd** [comment]
Removing the approved the lgtm labels since we still need QE and doc approvals.

**Comment by Vincent056** [comment]
> Removing the approved the lgtm labels since we still need QE and doc approvals.

QE has approved, waiting on the doc approve.

**Comment by rhmdnd** [comment]
Reapplying qe-approved for @xiaojiey 

/label qe-approved

**Review by sheriff-rh — COMMENTED** [review]
Looks good to me! Nice work Vincent.

**Review by rhmdnd — APPROVED** [review]
/lgtm
/approve
